zingolib/src/data.rs

```diff
@@ -3,60 +3,70 @@ pub mod proposal;
+    use zcash_address::ZcashAddress;
     use zcash_client_backend::zip321::Payment;
     use zcash_client_backend::zip321::TransactionRequest;
     use zcash_client_backend::zip321::Zip321Error;
-    use zcash_keys::address;
     use zcash_primitives::memo::MemoBytes;
     use zcash_primitives::transaction::components::amount::NonNegativeAmount;
 
     /// A list of Receivers
-    pub type Receivers = Vec<Receiver>;
+    pub type Destinations = Vec<Destination>;
 
     /// The superficial representation of the the consumer's intended receiver
     #[derive(Clone, Debug, PartialEq)]
-    pub struct Receiver {
-        pub(crate) recipient_address: address::Address,
+    pub struct Destination {
+        pub(crate) recipient: String,
         pub(crate) amount: NonNegativeAmount,
         pub(crate) memo: Option<MemoBytes>,
     }
-    impl Receiver {
+    impl Destination {
         /// Create a new Receiver
-        pub fn new(
-            recipient_address: address::Address,
-            amount: NonNegativeAmount,
-            memo: Option<MemoBytes>,
-        ) -> Self {
+        pub fn new(recipient: String, amount: NonNegativeAmount, memo: Option<MemoBytes>) -> Self {
             Self {
-                recipient_address,
+                recipient,
                 amount,
                 memo,
             }
         }
     }
-    impl From<Receiver> for Payment {
-        fn from(receiver: Receiver) -> Self {
-            Self {
-                recipient_address: receiver.recipient_address,
-                amount: receiver.amount,
-                memo: receiver.memo,
-                label: None,
-                message: None,
-                other_params: vec![],
-            }
-        }
+
+    /// anything that can go wrong parsing a TransactionRequest from a receiver
+    #[derive(thiserror::Error, Debug)]
+    pub enum DestinationParseError {
+        /// see Debug
+        #[error("Could not parse address: {0}")]
+        AddressParse(#[from] ParseError),
+        /// see Debug
+        #[error("Cant send memo to transparent receiver")]
+        MemoDisallowed,
+        /// see Debug
+        #[error("Could not build TransactionRequest: {0}")]
+        Request(#[from] Zip321Error),
     }
 
     /// Creates a [`zcash_client_backend::zip321::TransactionRequest`] from receivers.
-    pub fn transaction_request_from_receivers(
-        receivers: Receivers,
-    ) -> Result<TransactionRequest, Zip321Error> {
+    pub fn transaction_request_from_destinations(
+        receivers: Destinations,
+    ) -> Result<TransactionRequest, DestinationParseError> {
         let payments = receivers
             .into_iter()
-            .map(|receiver| receiver.into())
-            .collect();
+            .map(|receiver| {
+                Payment::new(
+                    ZcashAddress::try_from_encoded(receiver.recipient.as_str())
+                        .map_err(DestinationParseError::AddressParse)?,
```

Comment:
i dont understand. so much code has changed around parsing the address but it seems to have nothing to do with LRZ changes? it is being parsed here instead... why not just parse the address in the same place it was and leave everything alone?

Reply:
there are now 3 error cases for address parsing. memo-to-transparent error has been folded into create payment.